### ISSUE TITLE
EaserCAT-3000  PWM OUTPUT

### ISSUE THREAD

**Issue by TC227** [MetalMusings/MyOwnEtherCATDevice#14]
I'm trying to find the command component ///// Analog out
const byte DAC1_pin = PA4; to use as a spindle. I'm going to add an Op-amp
But I'm not sure if there's a command to generate PWM.

**Comment by MetalMusings** [comment]
It is Arduino platform so it is a basic analogWrite() command.
Bear in mind the pin output is only 0-3.3V so you may need additional circuitry to get +-10V

**Comment by TC227** [comment]
Do I need to enable it?
![err](https://github.com/user-attachments/assets/967864ce-8a9c-420a-ae5b-2faa09062dc8)

////// EtherCAT routines called regularly, to read data, do stuff and write data
void cb_set_outputs(void) // Get Master outputs, slave inputs, first operation
{
   for (int i = 0; i < min(sizeof(Obj.Outputs), sizeof(OUTPUTS)); i++)
      digitalWrite(OUTPUTS[i], Obj.Outputs[i] == 1 ? HIGH : LOW);
   // analogWrite(DAC1_pin, Obj.Voltage);
   // Encoder1.setLatch(Obj.IndexLatchEnable);
   // Encoder1.setScale(2000);

**Comment by TC227** [comment]
I edited main.cpp to link esi
![main](https://github.com/user-attachments/assets/e22e6e0c-e303-4c24-b1b0-ebeba1de2c7c)

and edited EEPROM generator to create soes-esi
![6007-7010](https://github.com/user-attachments/assets/e277fef1-b456-4e4c-b928-f3a5ced4f320)
![utype](https://github.com/user-attachments/assets/67310061-9038-479e-8050-90b2673db20a)

compiled and uploaded eeprom via twincat with newly added XML and it's done
![tw1](https://github.com/user-attachments/assets/9d6bf5e2-4b1a-43b8-a631-e81d1268ee41)

but twincat still doesn't find link to 6007 Voltage and 7010 IndexLatchEnable
what else do I need to change

**Comment by MetalMusings** [comment]
EEprom update
MCU update
ethercat-conf.xml update
hal update
all needs to be updated. For Twincat the eeprom and mcu needs to be updated.

**Comment by TC227** [comment]
Yes, it's funny that I forgot to update the MCU now that I've found it in LCNC.

**Comment by MetalMusings** [comment]
When using with a MCU, sometimes it is the content in the eeprom, sometimes the MCU's data model that are needed. Best to update both.

**Comment by TC227** [comment]
TWINCAT OK
![tw2ok](https://github.com/user-attachments/assets/9ed0b531-2211-4c33-850e-46dde12f7e76)

**Comment by MetalMusings** [comment]
There it is. But Voltage is an input. Don't you want that as an output?

**Comment by TC227** [comment]
I'm still confused, does the DAC voltage have to be moved to the output? And should the index be the input?

**Comment by TC227** [comment]
![9999](https://github.com/user-attachments/assets/ad22c246-109e-48f3-8376-2aaf9d5630aa)

**Comment by MetalMusings** [comment]
If you want to set a voltage or pwm duty cycle on the DAC pin, it should be set as output.
The terminology can be confusing. When doing the eeprom and looking at the soes code, input and output is from the Ethercat master's perspective.  Not from the client's. So yes it should be set as output.
Regarding inputs 1/0, start in twincat. See if it reacts when you set the pin high(3.3V) or low (gnd). If that works, do the same in linuxcnc. It can be anything from the hardware, collision with other functions like timers, or the setup.

**Comment by TC227** [comment]
Thanks for the advice, now I can control 0-3.3V at PA4, when I test 255 it outputs 3.3V.
I am now building the next OP-AMP, I will start experimenting with LM358.

**Comment by MetalMusings** [comment]
You can check out here to get more steps than 255, if you need that. Never tested myself.
https://www.stm32duino.com/viewtopic.php?t=1105

**Comment by DesertShedStudio** [comment]
@TC227 For a 3.3v spindle DAC circuit, you may find Peter's post here useful (Mesa 7i76 Spindle DAC Schematic) 
[https://forum.linuxcnc.org/39-pncconf/33991-configuration-7i92m-7i76x1d?start=40#110131](url)

**Comment by TC227** [comment]
Wow, that's right. So glad. I was looking for this post a few days ago but couldn't find it. I had a bad experience connecting +15 -15 which combined to 30V. I later found out that it should be 7.5+7.5 or 10+10. It damaged my 7i96 at 4001 and 4218. I still haven't fixed it to work. 

Thanks for helping me.
